### DISCUSSION TITLE
GlobalPlatformPro support Put Key AES SCP03 ?

### DISCUSSION THREAD

**Discussion by yp25** [martinpaljak/GlobalPlatformPro#343] [Q&A]
Hello, 
I would like to know whether GlobalPlatformPro support Put Key AES SCP03 ?

                    // Check for presence (thus replace)
                    // WORKAROUND: some cards reject the command if actually trying to replace existing key.
                    // List<GPKeyInfo> current = gp.getKeyInfoTemplate();
                    // boolean replace = current.stream().filter(p -> p.getVersion() == keyVersion).count() == 1 || args.has(OPT_REPLACE_KEY);
                    boolean replace = args.has(OPT_REPLACE_KEY);
                    if (kv.getPublic().isPresent()) {
                        gp.putKey(kv.getPublic().get(), keyVersion, replace);
                    } else if (kv.getSymmetric().isPresent()) {
                        byte[] k = kv.getSymmetric().get();
                        if (k.length != 16)
                            throw new IllegalArgumentException("Invalid key length: " + k.length);
                        // FIXME: implicit DES currently
                        gp.putKey(GPCrypto.des3key(k), keyVersion, replace);
                    } else {
                        throw new IllegalArgumentException("Only public and symmetric keys are supported for put-key");
                    }
                }`

**Comment by martinpaljak** [comment]
It does in master branch (to indicate that a 16 byte key is AES key not DES key, prepend key type `aes:4041...4F`. To put SCP03 keys, you still need to use the --lock options (which have supported SCP03 for many years)

**Reply by gwachob** [reply]
I've tried this with the following error
```shell
$ gp --lock aes:11111111111111111111111111111111                                                                                                                         
# Warning: no keys given, defaulting to 404142434445464748494A4B4C4D4E4F
Error: Cannot invoke "apdu4j.core.HexBytes.value()" because the return value of "joptsimple.OptionSet.valueOf(joptsimple.OptionSpec)" is null
```

I'm trying to set up scp03 keys on a card with a default dev SCP02 RSA keys in its ISD. 

Documentation says it should support SCP03.

I've also tried various things with `--put-key` but I'm getting 6A80 and I'm pretty sure the key-id being set is wrong but I honestly not sure how to get globalplatformpro to change the key-id i'm setting.

**Reply by martinpaljak** [reply]
I'd assume this is related to #383 and I expect to give it a look as soon as time permits...
